### PR TITLE
[IOPLT-295] Adding methods for Mongo CDC feature

### PR DESCRIPTION
**Pull request by drmarro**
#### List of Changes

Here's the list of changes made :
- Added all the utility methods for Mongo handling
- Added all the methods for Mongo CDC feature
- Implemented the generic interface with the methods created
- 
#### Motivation and Context

These methods have been added for handling the CDC with Mongo DB.

#### How Has This Been Tested?

Each method has been properly unit tested.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [X] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.

### CODE REVIEW

File: src/capturer/mongo/__tests__/utils.test.ts [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+  });
+  it("should get MongoDB collection", async () => {
+    jest
+      .spyOn(mockDb, "listCollections")
+      .mockReturnValueOnce(mockListCollectionsCursor);
+    jest
+      .spyOn(mockListCollectionsCursor, "toArray")
+      .mockReturnValueOnce(Promise.resolve(mockToArray));
```
Comment (AleDore) [inline src/capturer/mongo/__tests__/utils.test.ts:93]: we can move spyOn implementation outside single test an mock each implementation singularly :)

File: src/capturer/mongo/utils.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+//   pipe(
+//     TE.tryCatch(
+//       () => collection.find<T>({}).sort({ _id: -1 }).limit(1).tryNext(),
+//       (reason) =>
+//         new Error(
+//           `Unable to get the last inserted document from collection: " ${reason}`
+//         )
+//     ),
+//     TE.map(O.fromNullable)
+//   );
```
Comment (AleDore) [inline src/capturer/mongo/utils.ts:94]: why this function commented here?

Comment (drmarro) [reply src/capturer/mongo/utils.ts:94]: Sorry, this function is currently not used. I'll removed it.

File: src/capturer/mongo/__tests__/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+};
+describe("findDocumentByID", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  const findOneSpy = jest.spyOn(mockCollection, "findOne");
+  it("should get the document by ID", async () => {
+    findOneSpy.mockReturnValueOnce(Promise.resolve(mockWithIDDocument));
```
Comment (AleDore) [inline src/capturer/mongo/__tests__/utils.test.ts:250]: alternative
```suggestion
    findOneSpy.mockResolvedValueOnce(mockWithIDDocument);
```

File: src/capturer/mongo/__tests__/utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+const collectionSpy = jest.spyOn(mockDb, "collection");
+
+describe("getMongoCollection", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it("should get MongoDB collection", async () => {
+    listCollectionsSpy.mockReturnValueOnce(mockListCollectionsCursor);
+    toArraySpy.mockReturnValueOnce(Promise.resolve(mockToArray));
```
Comment (AleDore) [inline src/capturer/mongo/__tests__/utils.test.ts:93]:
```suggestion
    toArraySpy.mockResolvedValueOnce(mockToArray);
```

File: src/capturer/mongo/__tests__/utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    expect(mockDb.listCollections).toHaveBeenCalledWith({
+      name: collectionName,
+    });
+    expect(collectionSpy).toHaveBeenCalledWith(collectionName);
+    expect(result).toEqual(E.right(mockCollection));
+  });
+
+  it("should get empty MongoDB collection list ", async () => {
+    listCollectionsSpy.mockReturnValueOnce(mockListCollectionsCursor);
+    toArraySpy.mockReturnValueOnce(Promise.resolve([]));
```
Comment (AleDore) [inline src/capturer/mongo/__tests__/utils.test.ts:107]: ditto

File: src/capturer/mongo/__tests__/utils.test.ts [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+const createCollectionSpy = jest.spyOn(mockDb, "createCollection");
+
+describe("getOrCreateMongoCollection", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it("should get or create MongoDB collection", async () => {
+    listCollectionsSpy.mockReturnValueOnce(mockListCollectionsCursor);
+    toArraySpy.mockReturnValueOnce(Promise.resolve(mockToArray));
```
Comment (AleDore) [inline src/capturer/mongo/__tests__/utils.test.ts:150]: ditto

File: src/capturer/mongo/__tests__/utils.test.ts [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        new Error(
+          "Impossible to Get the testCollection collection: Error: Error while getting collection"
+        )
+      )
+    );
+  });
+
+  it("should create a MongoDB collection", async () => {
+    listCollectionsSpy.mockReturnValueOnce(mockListCollectionsCursor);
+    toArraySpy.mockReturnValueOnce(Promise.resolve([]));
```
Comment (AleDore) [inline src/capturer/mongo/__tests__/utils.test.ts:184]: ditto

File: src/capturer/mongo/__tests__/utils.test.ts [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+
+const closeSpy = jest.spyOn(mockClient, "close");
+
+describe("disconnectMongo", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+  });
+  it("should get MongoDB client disconnected", async () => {
+    closeSpy.mockReturnValueOnce(Promise.resolve(void 0));
```
Comment (AleDore) [inline src/capturer/mongo/__tests__/utils.test.ts:208]: ditto

File: src/capturer/mongo/__tests__/utils.test.ts [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+const document = { id: testID } as unknown as Document;
+const insertOneSpy = jest.spyOn(mockCollection, "insertOne");
+describe("insertDocument", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it("should insert the document", async () => {
+    insertOneSpy.mockReturnValueOnce(Promise.resolve(mockInsertOneResult));
```
Comment (AleDore) [inline src/capturer/mongo/__tests__/utils.test.ts:285]: ditto

File: src/factory/__tests__/cosmosDBService.test.ts [hunk 10]
```diff
@@ -1,17 +1,20 @@
...
+  getItemByID,
 } from "../../capturer/cosmos/utils";
 import { cosmosDBService } from "../cosmosDBService";
 
 const mockCosmosClient: CosmosClient = {} as CosmosClient;
 const mockDatabase: Database = {} as Database;
 const mockContainer: Container = {} as Container;
-
+const mockItem: ContinuationTokenItem = { id: "test" } as ContinuationTokenItem;
 jest.mock("../../capturer/cosmos/utils");
```
Comment (AleDore) [inline src/factory/__tests__/cosmosDBService.test.ts:18]: maybe we can use `spyOn` here?

File: src/factory/__tests__/mongoDBService.test.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  mongoConnect,
+} from "../../capturer/mongo/utils";
+import { mongoDBService } from "../mongoDBService";
+
+const mockMongoClient: MongoClient = {} as MongoClient;
+const mockDatabase: Db = {} as Db;
+const mockCollection: Collection = {} as Collection;
+const mockItem: ContinuationTokenItem = { id: "test" } as ContinuationTokenItem;
+
+jest.mock("../../capturer/mongo/utils");
```
Comment (AleDore) [inline src/factory/__tests__/mongoDBService.test.ts:19]: ditto , we can use `spyOn` and mock implementation on top of describes?

File: src/factory/__tests__/mongoDBService.test.ts [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+import { ContinuationTokenItem } from "../../capturer/cosmos/utils";
+import * as MongoUtils from "../../capturer/mongo/utils";
+import { mongoDBService } from "../mongoDBService";
+
+const mockMongoClient: MongoClient = {} as MongoClient;
+const mockDatabase: Db = {} as Db;
+const mockCollection: Collection = {} as Collection;
+const mockItem: ContinuationTokenItem = { id: "test" } as ContinuationTokenItem;
+
+jest.mock("../../capturer/mongo/utils");
```
Comment (AleDore) [inline src/factory/__tests__/mongoDBService.test.ts:15]: Can we remove this mock ?

File: src/factory/__tests__/mongoDBService.test.ts [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+      mockMongoClient,
+      "invalid-database"
+    );
+    expect(E.isLeft(result)).toBeTruthy();
+    expect(result).toEqual(left(new Error("Database error")));
+  });
+
+  it("should get resource successfully", async () => {
+    getMongoCollectionSpy.mockImplementationOnce(() =>
+      TE.rightTask(() => Promise.resolve(mockCollection))
```
Comment (AleDore) [inline src/factory/__tests__/mongoDBService.test.ts:69]: just as a suggestion we can safely change this expression with another one like:
```suggestion
      TE.right(mockCollection)()
```
or even
```suggestion
     T.of(mockCollection)
```

with `T` as `import * as T from "fp-ts/lib/Task"`

File: src/factory/__tests__/mongoDBService.test.ts (hunk 12 above)
Comment (drmarro) [reply src/factory/__tests__/mongoDBService.test.ts:15]: this command is necessary to spyOn methods. If not present, the spy will not work

**Review by AleDore — APPROVED** [review]

**Pull request merged**